contracts/DutchExchange.sol

```diff
@@ -507,41 +494,39 @@ contract DutchExchange {
         public
         returns (uint returned, uint tulipsIssued)
     {
-        // Requirements
         uint sellerBalance = sellerBalances[sellToken][buyToken][auctionIndex][user];
+        // R1
         require(sellerBalance > 0);
 
         // Get closing price for said auction
         fraction memory closingPrice = closingPrices[sellToken][buyToken][auctionIndex];
```

Comment:
I think `closingPrice` here can be declared with `storage` instead of `memory`.
`memory` creates a copy of the closing price in contract's storage; `storage` -- only a reference.

Same on lines 633, 674, 786, 787

Reply:
We used memory because it stores the struct only temporarily and is cheaper to use than storage